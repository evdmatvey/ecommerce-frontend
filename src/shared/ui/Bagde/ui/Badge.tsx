import clsx from 'clsx';
import { ReactNode } from 'react';
import styles from './Badge.module.css';

interface BadgeProps {
  color?: 'primary' | 'info' | 'danger' | 'warning' | 'success';
  variant?: 'solid' | 'translucent';
  iconPosition?: 'left' | 'only' | 'right';
  children?: ReactNode;
  icon?: ReactNode;
}

const Badge = ({
  variant = 'solid',
  color = 'primary',
  iconPosition = 'left',
  icon,
  children,
}: BadgeProps) => {
  const classes = clsx(styles.root, styles[color], {
    [styles.solid]: variant === 'solid',
    [styles.translucent]: variant === 'translucent',
    [styles.iconOnly]: iconPosition === 'only',
    [styles.icon]: icon,
  });

  return (
    <span className={classes} data-testid="badge-testid">
      {iconPosition === 'left' && (
        <span data-testid="icon-left-testid">{icon}</span>
      )}
      {iconPosition === 'only' ? (
        <span data-testid="icon-only-testid">{icon}</span>
      ) : (
        children
      )}
      {iconPosition === 'right' && (
        <span data-testid="icon-right-testid">{icon}</span>
      )}
    </span>
  );
};

export default Badge;
