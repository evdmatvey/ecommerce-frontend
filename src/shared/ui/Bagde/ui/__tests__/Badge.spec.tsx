import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import Badge from '../Badge';

export const badgeTestIds = {
  BADGE: 'badge-testid',
  ICON_LEFT: 'icon-left-testid',
  ICON_ONLY: 'icon-only-testid',
  ICON_RIGHT: 'icon-right-testid',
};

const TEST_TEXT = 'Budge!';
const TEST_ICON = <span>Icon</span>;

describe('Badge logic', () => {
  it('Should render badge correctly', () => {
    render(<Badge>{TEST_TEXT}</Badge>);
    const budge = screen.getByTestId(badgeTestIds.BADGE);
    expect(budge).toHaveTextContent(TEST_TEXT);
  });

  it('Should render badge correctly with left icon', () => {
    render(
      <Badge icon={TEST_ICON} iconPosition="left">
        {TEST_TEXT}
      </Badge>,
    );
    const budge = screen.getByTestId(badgeTestIds.BADGE);
    const leftIcon = screen.getByTestId(badgeTestIds.ICON_LEFT);
    expect(budge).toHaveTextContent(TEST_TEXT);
    expect(leftIcon).toBeInTheDocument();
  });

  it('Should render badge correctly with left icon (without position declaration)', () => {
    render(<Badge icon={TEST_ICON}>{TEST_TEXT}</Badge>);
    const budge = screen.getByTestId(badgeTestIds.BADGE);
    const leftIcon = screen.getByTestId(badgeTestIds.ICON_LEFT);
    expect(budge).toHaveTextContent(TEST_TEXT);
    expect(leftIcon).toBeInTheDocument();
  });

  it('Should render badge correctly with right icon', () => {
    render(
      <Badge icon={TEST_ICON} iconPosition="right">
        {TEST_TEXT}
      </Badge>,
    );
    const budge = screen.getByTestId(badgeTestIds.BADGE);
    const rightIcon = screen.getByTestId(badgeTestIds.ICON_RIGHT);
    expect(budge).toHaveTextContent(TEST_TEXT);
    expect(rightIcon).toBeInTheDocument();
  });

  it('Should render badge correctly with icon only', () => {
    render(<Badge icon={TEST_ICON} iconPosition="only" />);
    const icon = screen.getByTestId(badgeTestIds.ICON_ONLY);
    expect(icon).toBeInTheDocument();
  });
});

describe('Badge snapshots', () => {
  it('Badge with left icon (without position declaration)', () => {
    const { asFragment } = render(<Badge icon={TEST_ICON}>{TEST_TEXT}</Badge>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Badge with left icon', () => {
    const { asFragment } = render(
      <Badge icon={TEST_ICON} iconPosition="left">
        {TEST_TEXT}
      </Badge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Badge with right icon', () => {
    const { asFragment } = render(
      <Badge icon={TEST_ICON} iconPosition="right">
        {TEST_TEXT}
      </Badge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Badge with icon only', () => {
    const { asFragment } = render(
      <Badge icon={TEST_ICON} iconPosition="only" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
