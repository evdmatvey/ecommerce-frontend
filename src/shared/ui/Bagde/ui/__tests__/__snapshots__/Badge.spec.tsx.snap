// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge snapshots > Badge with icon only 1`] = `
<DocumentFragment>
  <span
    class="_root_5327bb _primary_5327bb _solid_5327bb _iconOnly_5327bb _icon_5327bb"
    data-testid="badge-testid"
  >
    <span
      data-testid="icon-only-testid"
    >
      <span>
        Icon
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Badge snapshots > Badge with left icon (without position declaration) 1`] = `
<DocumentFragment>
  <span
    class="_root_5327bb _primary_5327bb _solid_5327bb _icon_5327bb"
    data-testid="badge-testid"
  >
    <span
      data-testid="icon-left-testid"
    >
      <span>
        Icon
      </span>
    </span>
    Budge!
  </span>
</DocumentFragment>
`;

exports[`Badge snapshots > Badge with left icon 1`] = `
<DocumentFragment>
  <span
    class="_root_5327bb _primary_5327bb _solid_5327bb _icon_5327bb"
    data-testid="badge-testid"
  >
    <span
      data-testid="icon-left-testid"
    >
      <span>
        Icon
      </span>
    </span>
    Budge!
  </span>
</DocumentFragment>
`;

exports[`Badge snapshots > Badge with right icon 1`] = `
<DocumentFragment>
  <span
    class="_root_5327bb _primary_5327bb _solid_5327bb _icon_5327bb"
    data-testid="badge-testid"
  >
    Budge!
    <span
      data-testid="icon-right-testid"
    >
      <span>
        Icon
      </span>
    </span>
  </span>
</DocumentFragment>
`;
